refactor/cmake (#55)
TotalCross/totalcross: TotalCrossVM/CMakeLists.txt

> cmake_minimum_required(VERSION 3.13.5)

Reviewer: Has this version been validated? If we don't need 3.13, then we should reduce the requirement as we make it easier for people to build on their host distributions.

Reviewer 2: How so validated? We need these minimum version rquirement because we use some features of [cmake-generators(7) ](https://cmake.org/cmake/help/v3.13/manual/cmake-generators.7.html) that, theoretically, according to this documentation are part of CMake since version 3.13.x. The feature is related to [this doc](https://cmake.org/cmake/help/v3.13/generator/Visual%20Studio%209%202008.html). *And why is it relevant?* For me, it brings a slightly more *modular* experience, the generation for VS project became more similar to XCode generation. Both Win32 and WinCE (and their subprojects) are smoother to open into Visual Studio 2008 IDE.

Author: version required reduced to 3.5

# set the project name

> project(TCVM VERSION 1.0)

Reviewer: Wouldn't this be 7.0?

Reviewer 2: I don't know 😅, I think we can make this change.

> if(NOT DEFINED BUILD_SHARED_LIBS AND NOT DEFINED TCVM_SHARED)
>   if(CMAKE_GENERATOR STREQUAL Xcode)
>     set(TCVM_SHARED OFF) 
>     message("tcvm: ${TCVM_SHARED}")  
>   else() # others
>     set(TCVM_SHARED ON) 
>   endif()
> endif()

Reviewer: Take a look at the code suggestion below, it seems to be easier to understand and also provides an explanation for why this is required.

```suggestion
# By default we build the TCVM as a shared library, except for Xcode
# which we do it as a static library.
if(NOT DEFINED TCVM_SHARED)
  if(CMAKE_GENERATOR STREQUAL Xcode)
    set(TCVM_SHARED OFF)
  else()
    set(TCVM_SHARED ON)
  endif()
endif()

message("tcvm shared library is: ${TCVM_SHARED}")  
```

Reviewer 2: It will probably be one of the first changes after we get where we want (first merge with master).

# Find SDL2 and Skia

> set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/modules")

Reviewer: I'd move this to the beginning of the file, probably line 4 or 5, as this is generic code and also, this is not restricted to those modules but any modules you ever develop so the comment is misleading.

Reviewer 2: Sounds good to me.

# Find SDL2
IF (NOT (DEFINED ANDROID_ABI OR MSVC OR CMAKE_GENERATOR STREQUAL Xcode))
  find_package(SDL2 REQUIRED)
  include_directories(${SDL2_INCLUDE_DIR})
ENDIF (NOT (DEFINED ANDROID_ABI OR MSVC OR CMAKE_GENERATOR STREQUAL Xcode))

# Find Skia
if(NOT MSVC)
  find_package(Skia REQUIRED)
  include_directories(${SKIA_INCLUDE_DIRS})
endif(NOT MSVC)

> # Set C/CXX flags

Reviewer: This block fails here, so I think it should be dropped.

Reviewer 2: Can you give more details about the failure?

Author: fixed

if(UNIX)
  set(UNIX_C_FLAGS "-Wno-everything")
  set(UNIX_CXX_FLAGS "-Wno-everything")
endif(UNIX)

if(APPLE)

>   set(MACOS_C_FLAGS "-mmacosx-version-min=10.11")
>   set(MACOS_CXX_FLAGS "-mmacosx-version-min=10.11")

Reviewer: I think it makes it easier to find where we are changing those variables instead of adding an indirection for it.

```suggestion
  set(CMAKE_C_FLAGS  "${CMAKE_C_FLAGS} -mmacosx-version-min=10.11")
  set(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} -mmacosx-version-min=10.11")
```

Reviewer 2: We were thinking about code reuse, but as it is a small snippet and as we will not have many flags, it can really make reading easier. @flsobral what do you think about it?

Author: I agree, indirections are bad for readability.

Author: I changed the code to use generator expressions and modern cmake compile features. @acmlira is not a big fan of generator expressions, what do you think @otavio ?

Author: fixed

endif(APPLE)

> SET(CMAKE_C_FLAGS  "${CMAKE_C_FLAGS} ${UNIX_C_FLAGS} ${MACOS_C_FLAGS}")
> SET(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} ${UNIX_CXX_FLAGS} ${MACOS_CXX_FLAGS} -std=c++11")

Reviewer: ```suggestion
# We need to use C++ 11
set(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} -std=c++11")
```
